### PR TITLE
Task 3 Conformance checking

### PR DESCRIPTION
**Pull request by reichertli**

### CODE REVIEW

File: practical/ProcessMining/group2/conformance_checking/src/imported_pm4py/basic_playout.py
Comment (OhItsLena) [inline practical/ProcessMining/group2/conformance_checking/src/imported_pm4py/basic_playout.py:1]: Did you make any adjustments to the imported replay logic from pm4py? Otherwise I'm not sure what the benefit of having the code copied here instead of just imported from the package is.

File: practical/ProcessMining/group2/conformance_checking/src/comparison.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+
+com = Comparison()
+print(
+    com.model_2_model(
+        log_path,
+        [AlgoPm4Py.ALPHA, AlgoPm4Py.ALPHAPLUS, AlgoPm4Py.INDUCTIVEMINER],
+        2,
+    )
+)
```
Comment (OhItsLena) [inline practical/ProcessMining/group2/conformance_checking/src/comparison.py:234]: I would put experimental runs of the logic into a separate file or Jupyter notebook, so that the class files stay clean.

File: practical/ProcessMining/group2/conformance_checking/src/comparison.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+        Returns:
+        -------
+        List of tuples, where each tuple contains the name of the sublog and the conformance value.
+        """
+
+        def split_log_into_sublogs(log):
+            sublogs = []
+            # Initial split to divide the log into two equal parts
+            split_1, split_2 = split_train_test.split(log, train_percentage=0.5)
```
Comment (OhItsLena) [inline practical/ProcessMining/group2/conformance_checking/src/comparison.py:86]: From my understanding pm4py's `split_train_test.split()` does actually only shuffle and then randomly samples indices for the train and test log, which can easily be done without the package. If a stratified split is what you want, the scikit-learn `train_test_split()` has an implementation for that, but not easy to define what stratified would be for event log data anyway I guess.

File: practical/ProcessMining/group2/conformance_checking/tests/test_conformance_checking.py
Comment (OhItsLena) [inline practical/ProcessMining/group2/conformance_checking/tests/test_conformance_checking.py:1]: Maybe more tests that check for something else than 100 percent conformance would be more significant.

**Review by OhItsLena — COMMENTED** [review]
Good work. Interesting to see another approach to conformance implemented compared to our task. More type hints for the methods would make it easier to see what should go into them and what they return, as by now the project has quite a lot different custom types, pm4py imported types, etc.

File: practical/ProcessMining/group2/conformance_checking/src/check_conformance.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            inner_keys_2 = set(fpm_2.relations.get(key, {}).keys())
+            all_inner_keys = inner_keys_1.union(inner_keys_2)
+
+            for inner_key in all_inner_keys:
+                inner_v1 = fpm_1.relations.get(key, {}).get(inner_key, '')
+                inner_v2 = fpm_2.relations.get(key, {}).get(inner_key, '')
+                if inner_v1 == inner_v2:
+                    inner_dict_out[inner_key] = ''
+                else:
+                    inner_dict_out[inner_key] = '{}:{}'.format(inner_v1, inner_v2)
```
Comment (TJEWH) [inline practical/ProcessMining/group2/conformance_checking/src/check_conformance.py:48]: could be done by init inner_dict_out as defaultdict(str), so that you only need to handle` if inner_v1 != inner_v2`

File: practical/ProcessMining/group2/conformance_checking/src/comparison.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        sublogs = split_log_into_sublogs(log)
+
+        replayed_logs = []
+        # Run algorithm on each sublog
+        for i, sublog in enumerate(sublogs):
+            # Run pipeline
+            footprint_of_log, footprint_of_replayed_log = self.pipeline(
+                sublog, algorithm
+            )
+            visualize_sorted_dict(
```
Comment (TJEWH) [inline practical/ProcessMining/group2/conformance_checking/src/comparison.py:107]: For code quality, I wouldn't directly import external method, but import the parental class / file. Creates longer lines, but can immediately get the context of the method and distinguish it from nested functions e.g.

Concerns also other lines, I won't seperately reference

File: practical/ProcessMining/group2/conformance_checking/src/comparison.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                footprint_of_replayed_log.relations, "l2l_{}".format(i)
+            )
+            replayed_logs.append(footprint_of_replayed_log)
+
+        # Compare with original log
+        footprint_of_log, x = self.pipeline(log, algorithm)
+        visualize_sorted_dict(footprint_of_log.relations, "l2l_original")
+        comparison_values = []
+        conformance_checking = ConformanceChecking()
```
Comment (TJEWH) [inline practical/ProcessMining/group2/conformance_checking/src/comparison.py:116]: Arguable, if conformance checking needs it's own object that needs to be initialized instead of two functions `conformance.get_value` and `conformance.get_matrix`, object init is actually redundant.

File: practical/ProcessMining/group2/conformance_checking/src/generate_footprint.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            The relations dictionary to initialize the FootPrintMatrix.
+
+        Returns:
+        -------
+        FootPrintMatrix
+            A new FootPrintMatrix instance with the specified relations.
+        """
+        return cls(relations=relations)
+
+    def sort_fpm_rec(self, relations):
```
Comment (TJEWH) [inline practical/ProcessMining/group2/conformance_checking/src/generate_footprint.py:60]: contemplate renaming to sort_footprint_matrix_recursively, criptic method title's never solved anything ^^

**Review by TJEWH — COMMENTED** [review]
All in all, a convienient implementation. Just small remarks for code quality improvements from my side.
